File: app/src/main/java/ch/bailu/aat/services/sensor/internal/InternalSensorSDK23.java
package ch.bailu.aat.services.sensor.internal;

import android.content.Context;
import android.hardware.Sensor;
import android.hardware.SensorEventListener;
import android.hardware.SensorManager;

import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;

import ch.bailu.aat.services.sensor.Connector;
import ch.bailu.aat.services.sensor.SensorInterface;
import ch.bailu.aat.services.sensor.list.SensorItemState;
import ch.bailu.aat.services.sensor.list.SensorListItem;

@RequiresApi(api = 23)
public abstract class InternalSensorSDK23 implements SensorEventListener, SensorInterface {


    private final Context context;

    private final String name;
    private final String address;

    private boolean registered = false;

    private final Connector connector;
    private final SensorListItem item;


    public InternalSensorSDK23(Context c, SensorListItem i, Sensor sensor, int iid) {
        context = c;
        item = i;
        name = InternalSensorsSDK23.toName(sensor);
        address = InternalSensorsSDK23.toAddress(sensor);
        connector = new Connector(c, iid);

        if (item.lock(this)) {
            item.setState(SensorItemState.CONNECTING);
            item.setState(SensorItemState.CONNECTED);

            connector.connect();

            requestUpdates(this, sensor);

        }
    }

    /**
     * Was this instance successfully registered with the
     * SensorListItem?
     */
    protected final boolean isLocked() {
        return item.isLocked(this);
    }

    @Override
    public String getName() {
        return name;
    }


    @NonNull
    @Override
    public String toString() {
        return getName() + "@" + address + ":" + item.getSensorStateDescription(context);
    }



    @Override
    public void close() {
        if (item.unlock(this)) {

            connector.close();
            cancelUpdates(this);

            item.setState(SensorItemState.ENABLED);
        }
    }


    private void requestUpdates(SensorEventListener listener, Sensor sensor) {
        final SensorManager manager = context.getSystemService(SensorManager.class);

        if (manager != null) {
            manager.registerListener(listener, sensor, SensorManager.SENSOR_DELAY_NORMAL);
            registered = true;
        }
    }


    private void cancelUpdates(SensorEventListener listener) {
        final SensorManager manager = context.getSystemService(SensorManager.class);

        if (registered) {
            manager.unregisterListener(listener);
            registered = false;
        }
    }
}

File: aat-android/src/main/kotlin/ch/bailu/aat/services/sensor/internal/InternalSensorSDK23.kt
package ch.bailu.aat.services.sensor.internal

import android.content.Context
import android.hardware.Sensor
import android.hardware.SensorEventListener
import android.hardware.SensorManager
import androidx.annotation.RequiresApi
import ch.bailu.aat.services.sensor.Connector
import ch.bailu.aat.services.sensor.SensorInterface
import ch.bailu.aat.services.sensor.list.SensorItemState
import ch.bailu.aat.services.sensor.list.SensorListItem

@RequiresApi(api = 23)
abstract class InternalSensorSDK23(
    private val context: Context,
    private val item: SensorListItem,
    sensor: Sensor,
    iid: Int
) : SensorEventListener, SensorInterface {

    override val name: String = InternalSensorsSDK23.toName(sensor)
    private val address: String = InternalSensorsSDK23.toAddress(sensor)
    private var registered = false
    private val connector: Connector = Connector(context, iid)

    init {
        if (item.lock(this)) {
            item.state = SensorItemState.CONNECTING
            item.state = SensorItemState.CONNECTED
            connector.connect()
            requestUpdates(this, sensor)
        }
    }

    /**
     * Was this instance successfully registered with the
     * SensorListItem?
     */
    protected val isLocked: Boolean
        get() = item.isLocked(this)

    override fun toString(): String {
        return name + "@" + address + ":" + item.getSensorStateDescription(context)
    }

    override fun close() {
        if (item.unlock(this)) {
            connector.close()
            cancelUpdates(this)
            item.state = SensorItemState.ENABLED
        }
    }

    private fun requestUpdates(listener: SensorEventListener, sensor: Sensor) {
        val manager = context.getSystemService(SensorManager::class.java)
        if (manager != null) {
            manager.registerListener(listener, sensor, SensorManager.SENSOR_DELAY_NORMAL)
            registered = true
        }
    }

    private fun cancelUpdates(listener: SensorEventListener) {
        val manager = context.getSystemService(SensorManager::class.java)
        if (registered) {
            manager.unregisterListener(listener)
            registered = false
        }
    }
}
